File: blocket.se/worker.py
#!/usr/bin/env python3

import crawler
import json
import logging
import os
import pika
import sys

OFFERS_QUEUE = 'offers'
BLOCKET_SEARCHS_QUEUE = 'blocket_searchs'

def search_blocket(rabbit_channel):
    blocket = crawler.Crawler()

    for offer in blocket.offers():
        logging.info('Found offer "%s" -- %s', offer['title'], offer['identifier'])
        rabbit_channel.basic_publish(exchange='', routing_key=OFFERS_QUEUE, body=json.dumps(offer), properties=pika.BasicProperties(
            delivery_mode = 2, # make message persistent
        ))

def search_request_received(channel, method, properties, body):
    print('Received search request!')

    search_blocket(channel)

if __name__ == '__main__':
    logging.basicConfig(format='%(asctime)s - %(message)s', level=logging.INFO)
    rabbitMqUrl = os.environ.get('RABBITMQ_URL')

    if rabbitMqUrl is None:
        print("The DSN to use to connect to RabbitMq must be specified by the RABBITMQ_URL environment variable.", file=sys.stderr)
        sys.exit(1)

    rabbitmq = pika.BlockingConnection(pika.URLParameters(rabbitMqUrl))
    channel = rabbitmq.channel()
    channel.queue_declare(queue=OFFERS_QUEUE, durable=True)
    channel.queue_declare(queue=BLOCKET_SEARCHS_QUEUE, durable=True)

    logging.info('Waiting for search requests…')
    channel.basic_consume(search_request_received, queue=BLOCKET_SEARCHS_QUEUE)

    try:
        channel.start_consuming()
    except KeyboardInterrupt:
        rabbitmq.close()

File: blocket.se/worker.go
package main

import (
	"encoding/json"
	"os"

	log "github.com/sirupsen/logrus"
	"github.com/streadway/amqp"
)

const OffersQueue = "offers"
const BlocketSearchsQueue = "blocket_searchs"

func crawlOffers(channel *amqp.Channel) {
	resultChan, err := Crawl()
	if err != nil {
		log.Fatalf("Could not crawl offers: %s", err)
	}

	for offer := range resultChan {
		jsonOffer, err := json.Marshal(offer)
		if err != nil {
			log.Fatalf("Could not marshal offer: %s", err)
		}

		log.Infof("Found offer \"%s\" -- %s\n", offer.Title, offer.Identifier)

		err = channel.Publish(
			"",          // exchange
			OffersQueue, // routing key
			false,       // mandatory
			false,       // immediate
			amqp.Publishing{
				ContentType: "application/json",
				Body:        jsonOffer,
			},
		)
		if err != nil {
			log.Fatalf("Could not publish offer: %s", err)
		}
	}
}

func main() {
	rabbitMqUrl, ok := os.LookupEnv("RABBITMQ_URL")
	if !ok {
		log.Fatal("The DSN to use to connect to RabbitMq must be specified by the RABBITMQ_URL environment variable.")
	}

	rabbitConn, err := amqp.Dial(rabbitMqUrl)
	if err != nil {
		log.Fatalf("Could not connect to RabbitMq: %s", err)
	}
	defer rabbitConn.Close()

	channel, err := rabbitConn.Channel()
	if err != nil {
		log.Fatalf("Could not connect to RabbitMq channel: %s", err)
	}
	defer channel.Close()

	_, err = channel.QueueDeclare(
		OffersQueue, // name
		true,        // durable
		false,       // delete when unused
		false,       // exclusive
		false,       // no-wait
		nil,         // arguments
	)
	if err != nil {
		log.Fatalf("Could not declare RabbitMq offers queue: %s", err)
	}

	_, err = channel.QueueDeclare(
		BlocketSearchsQueue, // name
		true,                // durable
		false,               // delete when unused
		false,               // exclusive
		false,               // no-wait
		nil,                 // arguments
	)
	if err != nil {
		log.Fatalf("Could not declare RabbitMq search requests queue: %s", err)
	}

	msgs, err := channel.Consume(
		BlocketSearchsQueue, // queue
		"",                  // consumer
		true,                // auto-ack
		false,               // exclusive
		false,               // no-local
		false,               // no-wait
		nil,                 // args
	)
	if err != nil {
		log.Fatalf("Could not register search offers consumer: %s", err)
	}

	log.Printf("Waiting for search requests. To exit press CTRL+C")

	for _ = range msgs {
		go func() {
			log.Printf("Received search request!")

			crawlOffers(channel)
		}()
	}
}
